docs/Implementing-easyblocks.rst

```diff
@@ -0,0 +1,486 @@
+
+Specific aspects of easyblocks
+------------------------------
+
+
+.. _implementing_easyblocks_custom_parameters:
+
+Custom easyconfig parameters
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+In an easyblock, additional custom easyconfig parameters can be defined to steer the behaviour of the easyblock.
+This is done via the ``extra_options`` static method. Custom parameters can be defined to be mandatory or optional.
+
+The example below shows how this can be implemented:
+
+.. code:: python
+
+    from easybuild.easyblocks.generic.configuremake import ConfigureMake
+    from easybuild.framework.easyconfig import CUSTOM, MANDATORY
+
+    class EB_Example(ConfigureMake):
+        """Custom easyblock for Example"""
+
+        @staticmethod
+        def extra_options():
+            """Custom easyconfig parameters for Example"""
+            extra_vars = {
+                'required_example_param': [None, "Help text for required example custom parameter", MANDATORY],
+                'optional_example_param': [None, "Help text for (optional) example custom parameter", CUSTOM],
+            }
+            return ConfigureMake.extra_options(extra_vars)
+
+The first element in the list of a defined custom parameter corresponds to the default value for that parameter
+(both ``None`` in the example above).
+
+
+.. _implementing_easyblocks_constructor:
+
+Easyblock constructor
+~~~~~~~~~~~~~~~~~~~~~
+
+In the ``class`` constructor of the easyblock, i.e. the ``__init__`` method, one or more class variables
+can be initialised. These can be used for sharing information between different ``*_step`` methods in the easyblock.
+
+For example:
+
+.. code:: python
+
+    from easybuild.framework.easyblock import EasyBlock
+
+    class EB_Example(EasyBlock):
+        """Custom easyblock for Example"""
+
+        def __init__(self, *args, **kwargs):
+            """Constructor for Example easyblock, initialises class variables."""
+
+            # call out to original constructor first, so 'self' (i.e. the class instance) is initialised
+            super(EB_Example, self).__init__(*args, **kwargs)
+
+            # initialise class variables
+            self.var1 = None
+            self.var2 = None
+
+
+.. _implementing_easyblocks_files:
+
+Reading/writing/copying/patching files
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+File manipulation is a common use case for implementing easyblocks, hence the EasyBuild framework provides a
+number of useful functions related to this, including:
+
+* ``read_file(<path>)``: read file at a specified location and returns its contents
+
+* ``write_file(<path>, <text>)`` at a specified location with provided contents;
+  to append to an existing file, use ``append=True`` as an extra argument
+
+* ``copy_file(<src>, <dest>)`` to copy an existing file
+
+* ``apply_regex_substitutions(<path>, <list of regex substitions>)`` to patch an existing file
+
+All of these functions are provided by the ``easybuild.tools.filetools`` module.
```

Comment:
That's a nightmare to maintain, I wanted to focus on just the most significant ones.